### ISSUE TITLE
npm2rpm only intermittently continues beyond module unpacking

### ISSUE THREAD

**Issue by domcleal** [theforeman/npm2rpm#3]
npm2rpm is only sometimes getting passed the "Unpacking" stage and continuing onto the "Finished extracting" and further stages:

<pre>
# node_modules/.bin/npm2rpm -n jquery-flot -v 0.8.3 -s single
---- npm2rpm ----
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
 - Starting npm module download: http://registry.npmjs.org/jquery-flot/-/jquery-flot-0.8.3.tgz
 - Unpacking in /tmp/npm2rpm-3323nKvga3Mskca7 ...
# node_modules/.bin/npm2rpm -n jquery-flot -v 0.8.3 -s single
---- npm2rpm ----
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
 - Starting npm module download: http://registry.npmjs.org/jquery-flot/-/jquery-flot-0.8.3.tgz
 - Unpacking in /tmp/npm2rpm-3332453y4Wj7cpef ...
# node_modules/.bin/npm2rpm -n jquery-flot -v 0.8.3 -s single
---- npm2rpm ----
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
 - Starting npm module download: http://registry.npmjs.org/jquery-flot/-/jquery-flot-0.8.3.tgz
 - Unpacking in /tmp/npm2rpm-3341jdCgwP8fCohO ...
# node_modules/.bin/npm2rpm -n jquery-flot -v 0.8.3 -s single
---- npm2rpm ----
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
 - Starting npm module download: http://registry.npmjs.org/jquery-flot/-/jquery-flot-0.8.3.tgz
 - Unpacking in /tmp/npm2rpm-33504KF7EmJzfGo7 ...
 - Finished extracting for jquery-flot
 - Reading package.json for jquery-flot
 - Finished reading package.json for jquery-flot
 - Fetching flattened list of production dependencies
Cannot read property 'length' of undefined
</pre>

I have to run it multiple times to get the expected behaviour. The exit code is always zero.

**Comment by domcleal** [comment]
Possibly the finish event handler is only being added after the piping (in extractTar) from the request_stream to the directory is complete? It may need the error/finish event handlers to be registered before calling `request_stream.pipe(write_stream)`.

Edit: changing this didn't help

**Comment by dLobatog** [comment]
Hm... It appears to work for all other packages I've tried, I'll check why it struggles with this one

**Comment by dLobatog** [comment]
``` javascript
var totaldata = 0
tar_extract['stream'].on('data', (chunk) => {
  totaldata += chunk.length;
  console.log(`Received ${chunk.length} bytes of data. - total data ${totaldata}`);
})
```

Listening to this event shows that sometimes it does NOT download all the data needed. e.g: I got 499255, 403511, 475703, 373303...

Not sure what's causing it yet, but it's a start.

**Comment by dLobatog** [comment]
duh, found it thanks to - https://github.com/alanhoff/node-tar.gz/issues/40

The createWriteStream() stream we're getting is the one where the tar is finished extracting - not downloading. So if the tar finishes to extract before the file is fully downloaded, the program halts (and we cannot capture the 'on('error')' for that stream, that's why we didn't see any errors on the stream...

Fixed via d88ea7daf4b89723fc417db578bff129dd17ab11
